typing/typecore.ml

```diff
@@ -2994,6 +3040,132 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 sargs =
+              | Tmod_apply (p1, p2, _) ->
+                  Path.Papply(extract_path p1, extract_path p2)
+              | Tmod_constraint (p, _, _, _) ->
+                  extract_path p
+              | _ -> raise Not_found
+            in
+            let arg = Arg (Typed_arg { targ = texp }) in
+            begin
+              try
+                let path = extract_path me in
+                let ty_res =
+                    Option.value ~default:t
+                        (instance_funct ~id_in:(Ident.of_unscoped id)
+                                        ~p_out:path ~fixed:false t) in
+                let ty_res0 =
+                    Option.value ~default:t0
+                        (instance_funct ~id_in:(Ident.of_unscoped id0)
+                                        ~p_out:path ~fixed:false t0) in
+                loop ty_res ty_res0 ((l, arg) :: rev_args) remaining_sargs
+              with Not_found ->
+                try
+                  let env_t = Env.add_module (Ident.of_unscoped id) Mp_present
+                                              me.mod_type env in
+                  let env_t0 = Env.add_module (Ident.of_unscoped id0) Mp_present
+                                              me.mod_type env in
+                  identifier_escape env_t [id] t;
+                  identifier_escape env_t0 [id0] t0;
+                  loop t t0 ((l, arg) :: rev_args) remaining_sargs
+                with Unify trace ->
+                  raise (Error(sarg.pexp_loc, env,
+                        Cannot_infer_functor_path trace));
+            end
+          | Some _ | None ->
+            unify_to_arrows begin fun trace ->
+              match me_opt with
+              | Some (_, _, sarg) ->
+                  (* check that the type of the argument is indeed a package *)
+                  let _ =
+                    !type_argument' env sarg
+                      (newty2 ~level:(get_level ty_fun') (Tpackage (p, fl)))
+                      (newty2 ~level:(get_level ty_fun0) (Tpackage (p0, fl0)))
+                   in
+                  (* if it is a package of the expected type then we say that
+                     could not extract a path from it *)
+                  let previous_arg_loc = previous_arg_loc rev_args ~funct in
+                  let loc = Location.{
+                    loc_start = funct.exp_loc.loc_start;
+                    loc_end = previous_arg_loc.loc_end;
+                    loc_ghost = previous_arg_loc.loc_ghost
+                                  && funct.exp_loc.loc_ghost
+                  } in
+                  raise (Error(loc, env, Cannot_infer_functor_path trace))
+              | None ->
+                  let ty_res = remaining_function_type_for_error ty_fun rev_args
+                  in
+                  raise(Error(funct.exp_loc, env,
+                              Cannot_commute_label ty_res))
+            end;
+            loop ty_fun' ty_fun0 rev_args sargs
+          end
```

Comment:
I wish that most of the 126 lines you just added were moved to an auxiliary function, to keep the main function more readable. It might feel a bit ironic to make nitpick comments about good code structure in the middle of a file of more than 7K lines, but new code could strive to be better.